### ISSUE TITLE
Different batch size different result with inception_v2

### ISSUE THREAD

**Issue by JackieLeeTHU11** [tensorflow/tensorflow#11295]
I user inception_v2 as a base network for classification. During training, the batchsize=128. During testing, if the batchsize=128, everything is ok. However, if the batchsize is smaller than 128 the results are different. And the precision declines as the batchsize drops. If the batchsize=1, the network will failed. I also used inception_v3 and inception_v1, the same problems appered. However, if the base network is replaced with Alex network (tensorflow), everything goes well. 
I also replace the inception_v2 with vgg (slim), and everything goes well. 
The bug is associated with inception_v1~v3.  I think I have not used inception_v2 properly. Did anyone encounter similar problems?

**Comment by jubjamie** [comment]
_Warning: As this doesn't appear to be a bug with Tensorflow, the devs may ask for this to be moved to Stack Overflow._

It looks like you are asking about what Batch Size is. It is expected that reducing the batch size will reduce the accuracy. This is because you are giving your model a limited view of the data. But this is ok because we go over all the batches. There is a trade off with batch size and other performance aspects.

See this question for information https://stats.stackexchange.com/questions/164876/tradeoff-batch-size-vs-number-of-iterations-to-train-a-neural-network

**Comment by JackieLeeTHU11** [comment]
@jubjamie Hi Jubjamie, thanks for your reply. However, my question is not about how to select the batch size during training, but about the batch size during testing. I found the output are complete different when the batch size are different during testing.

**Comment by andydavis1** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by JackieLeeTHU11** [comment]
@andydavis1 Thanks, I will.

**Comment by js-instantor** [comment]
I am experiencing the exact same bug. I get different results depending on the batch size. Dropout is not the cause. Any ideas?

**Comment by linwangolo** [comment]
...

**Comment by clover978** [comment]
@JackieLeeTHU11 I meet the same problem. Have you solved it? Thanks

**Comment by azadef** [comment]
Do you update moving_mean and moving_variance variables in batch normalization during optimization?
I had the same problem with resnet and I added this in training, then it worked fine:
`update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)`
` with tf.control_dependencies(update_ops):`
    `train_op = opt.minimize(loss)`

**Comment by wincle** [comment]
@JackieLeeTHU11  Did you solve the problem?

**Comment by zuoanqh** [comment]
hmm, if its about batch size in **testing** it looks indeed a problem with tensorflow. why is this closed?

**Comment by azadef** [comment]
It's not a problem in tensorflow, just lacking of well documentation. Batch normalization variables are not updated in the graph automatically during optimization if you don't explicitly add them to dependency as I wrote in my earlier comment. That's why this only happens in inception and resnet. The other networks don't have batch norm as far as I know.

**Comment by xeoi** [comment]
@azadef It still doesn't work to me. Can you share me your script?

**Comment by xeoi** [comment]
I solved this problem. As @azadef said, the problem is because of batch norm. We should set is_training=True in training and False in testing. Also, use train_op = slim.learning.create_train_op(). instead of train_op = optimizer.minimize().
